File: mobile/pranathi-employee-android/app/src/main/java/com/pranathi/employeeapp/ApiClient.kt
package com.pranathi.employeeapp

import org.json.JSONArray
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

data class ApiResult(val statusCode: Int, val body: JSONObject)

class ApiException(message: String) : Exception(message)

class ApiClient(private val baseUrl: String) {
    fun signup(email: String, password: String): ApiResult {
        return sendJson(
            path = "/api/mobile/signup",
            method = "POST",
            payload = JSONObject()
                .put("email", email)
                .put("password", password),
        )
    }

    fun login(email: String, password: String): ApiResult {
        return sendJson(
            path = "/api/mobile/login",
            method = "POST",
            payload = JSONObject()
                .put("email", email)
                .put("password", password),
        )
    }

    fun me(token: String): ApiResult {
        return sendJson(path = "/api/mobile/me", method = "GET", token = token)
    }

    fun history(token: String): JSONArray {
        val result = sendJson(path = "/api/mobile/attendance/history", method = "GET", token = token)
        return result.body.optJSONArray("history") ?: JSONArray()
    }

    fun punch(token: String, action: String): ApiResult {
        return sendJson(
            path = "/api/mobile/attendance/punch",
            method = "POST",
            token = token,
            payload = JSONObject().put("action", action),
        )
    }

    fun logout(token: String): ApiResult {
        return sendJson(path = "/api/mobile/logout", method = "POST", token = token)
    }

    private fun sendJson(
        path: String,
        method: String,
        token: String? = null,
        payload: JSONObject? = null,
    ): ApiResult {
        val normalizedBaseUrl = baseUrl.trim().trimEnd('/')
        val connection = (URL("$normalizedBaseUrl$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 15000
            readTimeout = 15000
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Content-Type", "application/json")
            if (!token.isNullOrBlank()) {
                setRequestProperty("Authorization", "Bearer $token")
            }
            doInput = true
        }

        if (payload != null) {
            connection.doOutput = true
            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(payload.toString())
            }
        }

        val responseCode = connection.responseCode
        val stream = if (responseCode in 200..299) connection.inputStream else connection.errorStream
        val responseText = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
        val json = if (responseText.isBlank()) JSONObject() else JSONObject(responseText)

        if (responseCode !in 200..299) {
            val detail = json.optString("detail").ifBlank { "Request failed with status $responseCode." }
            throw ApiException(detail)
        }

        return ApiResult(responseCode, json)
    }
}

File: desktop/pranathi-employee-macos/Sources/APIClient.swift
import Foundation

struct APIClient {
    let baseURL: String

    func signup(email: String, password: String) async throws -> AuthResponse {
        try await send(
            path: "/api/mobile/signup",
            method: "POST",
            payload: ["email": email, "password": password],
            token: nil,
            responseType: AuthResponse.self
        )
    }

    func login(email: String, password: String) async throws -> AuthResponse {
        try await send(
            path: "/api/mobile/login",
            method: "POST",
            payload: ["email": email, "password": password],
            token: nil,
            responseType: AuthResponse.self
        )
    }

    func history(token: String) async throws -> [AttendanceEvent] {
        let response: AttendanceHistoryResponse = try await send(
            path: "/api/mobile/attendance/history",
            method: "GET",
            payload: Optional<EmptyPayload>.none,
            token: token,
            responseType: AttendanceHistoryResponse.self
        )
        return response.history
    }

    func punch(token: String, action: String) async throws -> String {
        let response: JSONMessageResponse = try await send(
            path: "/api/mobile/attendance/punch",
            method: "POST",
            payload: ["action": action],
            token: token,
            responseType: JSONMessageResponse.self
        )
        return response.message ?? "Attendance updated."
    }

    func logout(token: String) async throws {
        let _: JSONMessageResponse = try await send(
            path: "/api/mobile/logout",
            method: "POST",
            payload: Optional<EmptyPayload>.none,
            token: token,
            responseType: JSONMessageResponse.self
        )
    }

    private func send<T: Decodable, P: Encodable>(
        path: String,
        method: String,
        payload: P?,
        token: String?,
        responseType: T.Type
    ) async throws -> T {
        guard let url = URL(string: normalizedBaseURL + path) else {
            throw APIClientError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = 15
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let token, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        if let payload {
            request.httpBody = try JSONEncoder().encode(payload)
        }

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIClientError.invalidResponse
        }

        if !(200...299).contains(httpResponse.statusCode) {
            let detail = (try? JSONDecoder().decode(APIErrorResponse.self, from: data).detail) ?? "Request failed with status \(httpResponse.statusCode)."
            throw APIClientError.server(detail)
        }

        if T.self == JSONMessageResponse.self, data.isEmpty {
            return JSONMessageResponse(message: nil) as! T
        }

        return try JSONDecoder().decode(T.self, from: data)
    }

    private var normalizedBaseURL: String {
        baseURL.trimmingCharacters(in: .whitespacesAndNewlines).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    }
}

enum APIClientError: LocalizedError {
    case invalidURL
    case invalidResponse
    case server(String)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "The backend URL is invalid."
        case .invalidResponse:
            return "The backend returned an invalid response."
        case .server(let message):
            return message
        }
    }
}

private struct EmptyPayload: Encodable {}

private struct JSONMessageResponse: Codable {
    let message: String?
}
